libs/cosmos-sdk/baseapp/gasuseddb.go

```diff
@@ -40,3 +66,145 @@ func initDb() db.DB {
 	}
 	return db
 }
+
+func (h *HistoryGasUsedRecordDB) UpdateGasUsed(key []byte, gasUsed int64) {
+	h.latestGuMtx.Lock()
+	h.latestGu[string(key)] = append(h.latestGu[string(key)], gasUsed)
+	h.latestGuMtx.Unlock()
+}
+
+func (h *HistoryGasUsedRecordDB) GetHgu(key []byte) *HguRecord {
+	v, ok := h.cache.Get(string(key))
+	if ok {
+		return v.(*HguRecord)
+	}
+	var record HguRecord
+	data, err := h.guDB.Get(key)
+	if err != nil || len(data) == 0 {
+		return nil
+	}
+	err = proto.Unmarshal(data, &record)
+	if err != nil {
+		return nil
+	}
+	return &record
+}
+
+func (h *HistoryGasUsedRecordDB) FlushHgu() {
+	if len(h.latestGu) == 0 {
+		return
+	}
+	latestMeanGu := make([]gasKey, len(h.latestGu))
+	for key, gus := range h.latestGu {
+		latestMeanGu = append(latestMeanGu, gasKey{
+			gas: meanInt64(gus...),
+			key: []byte(key),
+		})
+		delete(h.latestGu, key)
+	}
+	h.jobQueue <- func() { h.flushHgu(latestMeanGu...) } // closure function
+}
+
+func (h *HistoryGasUsedRecordDB) flushHgu(gks ...gasKey) {
+	for _, gk := range gks {
+		meanGu := gk.gas
+		key := gk.key
+		hgu := h.GetHgu(key)
+		if hgu == nil {
+			hgu = &HguRecord{
+				MaxGas:       meanGu,
+				LastBlockGas: meanGu,
+				HighGas:      meanGu,
+				StandardGas:  meanGu,
+			}
+		} else {
+			hgu.LastBlockGas = meanGu
+			if meanGu > hgu.MaxGas {
+				hgu.MaxGas = meanGu
+			}
+			if meanGu >= hgu.StandardGas*2 {
+				// abnormal cases which meanGu of last block is much higher than StandardGas
+				hgu.HighGas = int64(GasUsedFactor*float64(meanGu) + (1-GasUsedFactor)*float64(hgu.HighGas))
```

Comment:
has not update StandardGas, if meanGu keep hight every block, the meanGu will keep hight